File: exercism/python/wordy/wordy.py
from operator import add, sub, mul, floordiv
from functools import reduce

OPS = {'plus': add, 'minus': sub, 'divided': floordiv, 'multiplied': mul}


def calculate(question):
    initial, ops = tokenize(question)
    return reduce(lambda a, x: x[0](a, x[1]), ops, initial)


def tokenize(question):
    words = [w for w in question[8:-1].split(' ') if w != 'by']
    try:
        rest = [(OPS[op], int(n)) for op, n in zip(words[1::2], words[2::2])]
    except KeyError:
        raise ValueError
    return int(words[0]), rest

File: exercism/javascript/wordy/wordy.js
const fp = require('lodash/fp')

const operation = {
  plus: fp.add,
  minus: fp.subtract,
  divided: fp.divide,
  multiplied: fp.multiply
}

const toToken = (str) =>
  Number(str) || operation[str]

const tokenize = (phrase) =>
  fp.compact(phrase.split(' ').map(toToken))

const foldIn = (a, [f, b]) => f(a, b)

const evaluate = (tokens) =>
  fp.chunk(2, fp.tail(tokens)).reduce(foldIn, tokens[0])

class WordProblem {
  constructor (question) {
    this.tokens = tokenize(question.replace('?', ''))
  }
  answer () {
    if (this.tokens.length < 3) throw new ArgumentError()
    return evaluate(this.tokens)
  }
}

class ArgumentError extends Error {}

module.exports = { WordProblem, ArgumentError }
